### PR TITLE
WIP osd,librados: add dmclock abilities to librados allowing clients to specify QoS

### PR DESCRIPTION
**Pull request by ivancich**
This is a work in progress.

### CODE REVIEW

File: src/osd/mClockClientProfileQueue.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+      // construct a request wrapper
+      explicit RequestProxyWrapper(Request&& _request) :
+	request(new Request(std::move(_request))),
+	is_proxy(false)
+      {}
+
+      RequestProxyWrapper(RequestProxyWrapper&& _wrapper) {
+	std::swap(request, _wrapper.request);
+	std::swap(is_proxy, _wrapper.is_proxy);
```
Comment (cbodley) [inline src/osd/mClockClientProfileQueue.h]: `is_proxy` is uninitialized in this constructor

the compiler will generate these move operations for you though, so you should be safe to remove them altogether? see https://godbolt.org/g/zqZUtz for a working example

**Comment by cbodley** [comment]
looking at this from the perspective of radosgw as the librados client, it would be cool to create some set of profiles that could be shared between all radosgw instances

as an example use case, say we wanted to provide qos for our gc requests to make sure that garbage collection keeps up with our normal client traffic. we'd want these profiles to be global, so that there'd still be fairness even if one gateway was issuing nothing but gc requests, and another gateway was doing only client requests

would that be possible with this design? it seems like it would require named profiles, where the librados client would say "create this profile named 'radosgw-gc', or give me its id if it already exists"

**Comment by cbodley** [comment]
please don't kill this, friendly robot!

**Comment by ivancich** [comment]
Does adding a comment count as activity?

**Comment by mattbenjamin** [comment]
@ivancich this needs a rebase

**Comment by mattbenjamin** [comment]
unstale me my man

**Comment by mattbenjamin** [comment]
@ivancich I thought this merged...

**Comment by ivancich** [comment]
@mattbenjamin I handed it off to @mkogan1 many months ago, so I could focus on other stuff. But I think he's been focusing on other stuff as well.

**Comment by mattbenjamin** [comment]
unstale it

**Comment by ivancich** [comment]
freshen this

**Comment by liyichao** [comment]
@ivancich When will this pr be merged?

**Comment by ivancich** [comment]
@liyichao: @athanatos has expressed an interest in resurrecting this PR. I expect it would need a significant rebase, and probably some finishing up. It was not my intention to not use rho and delta, so that's oversight or mistake.

**Comment by athanatos** [comment]
@ivancich I've got the rebase done (probably) and am working on getting cbt to a point where I can test the efficacy.  https://github.com/athanatos/ceph/tree/sjust/wip-dmclock-librados-rebase

**Comment by liyichao** [comment]
@athanatos

**Comment by athanatos** [comment]
@ivancich @liyichao  Here's what I've got so far: https://github.com/athanatos/ceph/tree/sjust/wip-dmclock-cleanup

I narrowed the osd/opqueue interface to eliminate the _strict concept and created what I think is a simpler wrapper around the core dmclock implementation.  Would it make sense at this point to eliminate the other two mclock queues?  I'd like to address the background/client mix by adjusting the client provided values upon arrival and periodically thereafter rather than by using a multi-level queue and avoid the entire priority queue construct.

Additionally, I think I'd like to drop the librados component for the moment (though not the objecter component).  My rational is that having two (let's say) rbd volumes on the same client with vastly different mclock profiles could be problematic as there isn't any support upstream of the client->osd channel for prioritizing the sending of ops.  Until we choose to create that kind of machinery, we can always choose instead to create two clients?

Let me know what you think.

**Pull request closed without merging**